cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/shortestpath/SingleSourceShortestPath.scala

```diff
@@ -0,0 +1,44 @@
+package com.twitter.cassovary.algorithms.shortestpath
+
+import com.twitter.cassovary.graph.{Node, GraphDir, BreadthFirstTraverser, DirectedGraph}
+
+case class SingleSourceShortestPath(graph: DirectedGraph[Node], source: Int) extends ShortestPath {
+
+  val bfs = new BreadthFirstTraverser(graph, GraphDir.OutDir, Seq(source))
+  private val (nodes, depths)  = (bfs.toList, bfs.allDepths)
+  private val previous = nodes.map { n =>
+    n.id -> n.inboundNodes().filter { in => depths(in) == depths(n.id) - 1}
+  }.toMap
+
+  private def walkOnGraph(currStack: Stack, currTop: Int): (Stack, Int, Path) = {
+    val node  = currStack(currTop)(0)
+    val index = currStack(currTop)(1)
+
+    val path = if (node == source)
+      currStack.take(currTop + 1).reverse.map { k => k(0) }.toSeq
+    else
+      Seq.empty[Int]
+
+    val (stack, top) = if (previous(node).length > index) {
+      val newStack = if (currTop + 1 == currStack.length)
+        currStack :+ Seq(previous(node)(index), 0)
+      else
+        currStack.zipWithIndex.map { case (seq, in) => if (in == currTop + 1) Seq(previous(node)(index), 0) else seq}
+      (newStack, currTop + 1)
+    }
+    else {
+      val newStack = currStack.zipWithIndex.map { case (seq, in) => if (in == currTop - 1) Seq(seq(0), seq(1) + 1) else seq}
+      (newStack, currTop - 1)
+    }
+    (stack, top, path)
+  }
+
+  def shortestPaths(target: Int): Paths = {
+    lazy val stream: Stream[(Stack, Int, Path)] =
+      (List(Seq(target, 0)), 0, Seq.empty[Int]) #:: stream.map { case (s, t, p) => walkOnGraph(s,t) }
+
+    stream.takeWhile{ case (_, t, _) => t >= 0}.toList
+      .filter{ case (_, _, p) => p.nonEmpty }
+      .map{ _._3 }.toSeq
```

Comment:
`walkonGraph()` is probably the most opaque piece of code here.  Essentially, I am trying to efficiently determine _all_ shortest paths from node i to node j--not just _a_ shortest path from node i to node j.  

Here is the output from the stack manipulation when determining the paths from node 10 to node 14 in `TestGraphs.g6`.  The result we are after is a set of paths [(10,11,14), (10,12,14), (10,13,14)].  The first line is the state of the current stack, the second line is the value of current "top"--which is essentially how far we have travelled along a given path (which could include traveling backward).
```scala
List(List(14, 0))
0
List(List(14, 0), List(11, 0))
1
List(List(14, 0), List(11, 0), List(10, 0))
2
List(List(14, 0), List(11, 1), List(10, 0))
1
List(List(14, 1), List(11, 1), List(10, 0))
0
List(List(14, 1), List(12, 0), List(10, 0))
1
List(List(14, 1), List(12, 0), List(10, 0))
2
List(List(14, 1), List(12, 1), List(10, 0))
1
List(List(14, 2), List(12, 1), List(10, 0))
0
List(List(14, 2), List(13, 0), List(10, 0))
1
List(List(14, 2), List(13, 0), List(10, 0))
2
List(List(14, 2), List(13, 1), List(10, 0))
1
List(List(14, 3), List(13, 1), List(10, 0))
0
```
So, we keep track of a list of tuples(node, index).  Index is probably a bad name for the second part as it really keeps track of the number of shortest paths that pass through that given node.  So, we keep walking starting at target and ending at source and keep visiting every node.  And at each node we visit, we ask "Is there another node that will lead me to source?"  If so, we follow that path, if not, we walk backward further until we do reach a node that could lead to another shortest path.  

Just to be illustrative, here is the output for all shortest paths from node 10 to node 15.
```scala
List(List(15, 0))
0
List(List(15, 0), List(14, 0))
1
List(List(15, 0), List(14, 0), List(11, 0))
2
List(List(15, 0), List(14, 0), List(11, 0), List(10, 0))
3
List(List(15, 0), List(14, 0), List(11, 1), List(10, 0))
2
List(List(15, 0), List(14, 1), List(11, 1), List(10, 0))
1
List(List(15, 0), List(14, 1), List(12, 0), List(10, 0))
2
List(List(15, 0), List(14, 1), List(12, 0), List(10, 0))
3
List(List(15, 0), List(14, 1), List(12, 1), List(10, 0))
2
List(List(15, 0), List(14, 2), List(12, 1), List(10, 0))
1
List(List(15, 0), List(14, 2), List(13, 0), List(10, 0))
2
List(List(15, 0), List(14, 2), List(13, 0), List(10, 0))
3
List(List(15, 0), List(14, 2), List(13, 1), List(10, 0))
2
List(List(15, 0), List(14, 3), List(13, 1), List(10, 0))
1
List(List(15, 1), List(14, 3), List(13, 1), List(10, 0))
0
```
As you can see here, we only "walk" back to 15 once since we only need to walk back to 14 before we find another node that will lead to a shortest path.  I often find this output to be the most helpful in understanding why and how this algorithm works.  As per usual, I followed the example from networkx for shortest paths.  I will try to find the source.